filter commits in git.Commits based on sub-repo permissions

mollylogue opened the pull request:
This PR adds sub-repo permissions filtering on the `git.Commit` function by passing the `--name-only` flag when running `git log` to get the list of files modified in each commit along with the rest of the commit information. I decided to put this part up for review before implementing/testing the filtering for the other functions in this package so that I could get a signoff on this approach (or feedback on how to do it better) before moving forward. 

https://github.com/sourcegraph/sourcegraph/issues/27498

<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->
<!-- Please notify @delivery if this PR contains changes to CI that may need to be cherry-picked on to patch release branches -->

ryanslade commented on internal/vcs/git/commits_test.go:
```diff
@@ -314,26 +316,80 @@ func TestRepository_Commits(t *testing.T) {
+
+	tests := map[string]struct {
+		repo        api.RepoName
+		opt         CommitsOptions
```
Don't seem to use this

mollylogue replied:
Ah yeah that was for potential future testing.

ryanslade commented on internal/vcs/git/commits.go:
```diff
@@ -295,6 +354,9 @@ func commitLogArgs(initialArgs []string, opt CommitsOptions) (args []string, err
 	if opt.Path != "" {
 		args = append(args, "--", opt.Path)
 	}
+	if opt.NameOnly {
```
Can you double check that it works if they've added a non-empty path? I'm not sure how a `--` with no name parameter will play with more parameters following it.

mollylogue replied:
So, it does work, but I needed to specify the path last and the `--name-only` flag with the rest of the flags beforehand. I've added tests to demonstrate this behavior.

ryanslade reviewed (approved):
LGTM

I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit

mollylogue commented:
> LGTM
> 
> I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit

I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.

mollylogue commented:
@ryanslade I believe I've addressed your comments thus far. I've added to the test cases to have better coverage and validation (specifically around the case you mentioned w/ both `--read-only` and a path specified). One outstanding question is whether or not the default client has been initialized already for all of these use cases. I'm not entirely sure about that.

ryanslade commented:
> @ryanslade I believe I've addressed your comments thus far. I've added to the test cases to have better coverage and validation (specifically around the case you mentioned w/ both `--read-only` and a path specified). One outstanding question is whether or not the default client has been initialized already for all of these use cases. I'm not entirely sure about that.

Great, LGTM.

I'm not too worried about whether we've initialised everything yet as that will come out during testing and it doesn't matter yet as it's all disabled in prod anyway.

ryanslade commented:
> > LGTM
> > I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit
> 
> I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.

I just tested locally running `git log` against our sourcegraph repo and adding `--name-only` is significantly slower so I think your approach is worth it.

```
Benchmark 1: git log --name-only > blah.txt
  Time (mean ± σ):      2.102 s ±  0.057 s    [User: 1.929 s, System: 0.156 s]
  Range (min … max):    2.030 s …  2.182 s    10 runs

Benchmark 1: git log > blah.txt
  Time (mean ± σ):     277.9 ms ±   5.0 ms    [User: 249.2 ms, System: 25.3 ms]
  Range (min … max):   269.3 ms … 286.2 ms    10 runs
```

sashaostrikov reviewed (approved):
LGTM!

mollylogue commented:
> > > LGTM
> > > I wonder how bad the performance would be to always return file names? It may simplify the code quite a bit
> > 
> > 
> > I was wondering about that too, but I wanted to make the change as low-impact as possible. I'm open to making it the default, though.
> 
> I just tested locally running `git log` against our sourcegraph repo and adding `--name-only` is significantly slower so I think your approach is worth it.
> 
> ```
> Benchmark 1: git log --name-only > blah.txt
>   Time (mean ± σ):      2.102 s ±  0.057 s    [User: 1.929 s, System: 0.156 s]
>   Range (min … max):    2.030 s …  2.182 s    10 runs
> 
> Benchmark 1: git log > blah.txt
>   Time (mean ± σ):     277.9 ms ±   5.0 ms    [User: 249.2 ms, System: 25.3 ms]
>   Range (min … max):   269.3 ms … 286.2 ms    10 runs
> ```

Oof. ☹️  I really wish that I could've figured out a way to do the filtering directly in the path field (i.e. https://stackoverflow.com/questions/5685007/making-git-log-ignore-changes-for-certain-paths) but it didn't seem like the format was compatible with the path excludes values in the sub-repo permissions (and sub-repo permissions can have multiple path excludes rules). That would've probably been the most performant way to do this, but oh well 🤷‍♀️
